### PR TITLE
Filtered table

### PR DESCRIPTION
**Pull request by spiresheep**

### CODE REVIEW

File: source/filtered_table_model.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+  from './operations';
+import { TableModel } from './table_model';
+
+
+export abstract class Predicate {
+
+  /**
+   * @param row - The row to apply the predicate to.
+   */
+  public abstract applyPredicate(row: number, model: TableModel): boolean;
```
Comment (kranar) [inline source/filtered_table_model.ts]: Make this just a function instead of an abstract class.

File: source/filtered_table_model.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    this.operations = [];
+    this.dispatcher = new Kola.Dispatcher<Operation[]>();
+    this.model.connect(this.handleOperations.bind(this));
+  }
+
+  /** Marks the beginning of a transaction. In cases where a transaction is
+   *  already being processed, then the sub-transaction gets consolidated into
+   *  the parent transaction.
+   */
+  public beginTransaction(): void {
```
Comment (kranar) [inline source/filtered_table_model.ts]: These methods should not be public.

File: source/filtered_table_model.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+      array.push(this.model.get(rowIndex, i));
+    }
+    const row = new ArrayTableModel();
+    row.addRow(array);
+    return row;
+  }
+
+  private rowAdded(operation: AddRowOperation) {
+    const rowAddedIndex = operation.index;
+    const truthyness =
```
Comment (kranar) [inline source/filtered_table_model.ts]: ```const isFiltered = ...```

File: tests/filtered_table_model_tester.ts [hunk 4]
```diff
@@ -0,0 +1,720 @@
+import { Expect, Test } from 'alsatian';
+import { AddRowOperation, ArrayTableModel, FilteredTableModel, Operation,
+  Predicate, RemoveRowOperation, TableModel, UpdateValueOperation }
+  from '../source';
+
+class MockPredicate implements Predicate {
+
```
Comment (kranar) [inline tests/filtered_table_model_tester.ts]: Remove blank line on line 7.

File: tests/filtered_table_model_tester.ts [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+  public applyPredicate(row: number, model: TableModel): boolean {
+    if(model.get(row, 0) >= 0) {
```
Comment (kranar) [inline tests/filtered_table_model_tester.ts]: ```return model.get(row, 0) >= 0;```

**Review by kranar — CHANGES_REQUESTED** [review]
This looks excellent.

File: source/filtered_table_model.ts (hunk 1 above)
Comment (spiresheep) [reply source/filtered_table_model.ts]: Should it be a currying function?

**Comment by kranar** [comment]
Nope.

File: source/filtered_table_model.ts [hunk 6]
```diff
@@ -4,13 +4,12 @@ import { AddRowOperation, Operation, RemoveRowOperation, UpdateValueOperation }
...
-
-  /**
-   * @param row - The row to apply the predicate to.
-   */
-  public abstract applyPredicate(row: number, model: TableModel): boolean;
+/**
+ * @param row - The row to apply the predicate to.
+ * @param model - The model that contains the row.
+ */
+export function applyPredicate(row: number, model: TableModel): boolean {
```
Comment (kranar) [inline source/filtered_table_model.ts]: Remove this function entirely.

File: source/filtered_table_model.ts [hunk 7]
```diff
@@ -19,10 +18,15 @@ export class FilteredTableModel extends TableModel {
    * @param source - The underlying model to filter.
    * @param predicate - The predicate used to filter the table.
    */
-  public constructor(model: TableModel, predicate: Predicate) {
+  public constructor(model: TableModel,
+      predicate?: (row: number, model: TableModel) =>  boolean) {
```
Comment (kranar) [inline source/filtered_table_model.ts]: The predicate is mandatory.

File: source/filtered_table_model.ts [hunk 8]
```diff
@@ -35,15 +39,15 @@ export class FilteredTableModel extends TableModel {
    *  already being processed, then the sub-transaction gets consolidated into
    *  the parent transaction.
    */
-  public beginTransaction(): void {
+  private beginTransaction(): void {
```
Comment (kranar) [inline source/filtered_table_model.ts]: Private methods are not documented, are placed after all public methods, and generally omit the return type.

File: source/filtered_table_model.ts [hunk 9]
```diff
@@ -113,9 +117,9 @@ export class FilteredTableModel extends TableModel {
 
   private rowAdded(operation: AddRowOperation) {
     const rowAddedIndex = operation.index;
-    const truthyness =
-      this.predicate.applyPredicate(0, operation.row);
-    if(truthyness) {
+    const isFiltered =
+      this.predicate(0, operation.row);
```
Comment (kranar) [inline source/filtered_table_model.ts]: No need to split the line.

File: source/filtered_table_model.ts [hunk 10]
```diff
@@ -113,9 +117,9 @@ export class FilteredTableModel extends TableModel {
...
+    if(isFiltered) {
```
Comment (kranar) [inline source/filtered_table_model.ts]: ```isFiltered``` is only ever read from once, no need to have a variable for this at all.

File: source/filtered_table_model.ts (hunk 7 above)
Comment (kranar) [inline source/filtered_table_model.ts]: The predicate should be ```(model: TableModel, row: number)```. Generally if one parameter depends on another, then it is listed afterwards. In this case the ```row``` depends on the ```model```.

**Review by kranar — CHANGES_REQUESTED** [review]

File: tests/filtered_table_model_tester.ts [hunk 11]
```diff
@@ -4,7 +4,7 @@ import { AddRowOperation, ArrayTableModel, FilteredTableModel, Operation,
   from '../source';
 
 class MockPredicate {
-  public applyPredicate(row: number, model: TableModel): boolean {
+  public applyPredicate(model: TableModel, row: number): boolean {
```
Comment (kranar) [inline tests/filtered_table_model_tester.ts]: Remove this class, you can define a function if you'd like but no need to have it be part of a class.

File: tests/filtered_table_model_tester.ts [hunk 12]
```diff
@@ -3,10 +3,8 @@ import { AddRowOperation, ArrayTableModel, FilteredTableModel, Operation,
   RemoveRowOperation, TableModel, UpdateValueOperation }
   from '../source';
 
-class MockPredicate {
-  public applyPredicate(model: TableModel, row: number): boolean {
-    return model.get(row, 0) >= 0;
-  }
+function applyPredicate(model: TableModel, row: number): boolean {
+  return (model.get(row, 0) >= 0);
```
Comment (kranar) [inline tests/filtered_table_model_tester.ts]: No brackets around the return.

**Review by kranar — APPROVED** [review]

**Pull request merged**